### ISSUE TITLE
JUnit throws SIGBUSes when used from c++?

### ISSUE THREAD

**Issue by tallavi** [junit-team/junit5#2132]
Hi!

I'm using CLion with a c++ project (cmake), which starts a jvm. The java part is built with gradle. The project works, but I'm having a problem with debugging.

When I start the JVM, I immediately get a SIGSEGV. I understand that it's normal, it comes from the jvm itself, and there's no workaround except ignoring SIGSEGV. A bit annoying but not too bad as it only happens once per session.

BUT, after that, I continue debugging, and I get constant SIGBUS signals with this stack trace:

```
<unknown> 0x000000011f108385
<unknown> 0x000000011761dca7
<unknown> 0x000000011761dca7
<unknown> 0x000000011761da00
<unknown> 0x000000011761da00
<unknown> 0x000000011761da00
<unknown> 0x000000011761da00
<unknown> 0x000000011761dae2
<unknown> 0x000000011761da00
<unknown> 0x000000011761dae2
<unknown> 0x000000011761da00
<unknown> 0x000000011761dae2
<unknown> 0x000000011761da00
<unknown> 0x0000000117614849
JavaCalls::call_helper(JavaValue*, methodHandle const&, JavaCallArguments*, Thread*) 0x000000010bf3a582
StackWalk::fetchFirstBatch(BaseFrameStream&, Handle, long, int, int, int, objArrayHandle, Thread*) 0x000000010c227cac
StackWalk::walk(Handle, long, int, int, int, objArrayHandle, Thread*) 0x000000010c2278fc
JVM_CallStackWalk 0x000000010bfb14a2
<unknown> 0x0000000117623950
<unknown> 0x000000011761da00
<unknown> 0x000000011761da00
<unknown> 0x000000011761da00
<unknown> 0x000000011761da00
<unknown> 0x000000011761da00
<unknown> 0x000000011761da00
<unknown> 0x000000011761da00
<unknown> 0x000000011761da00
<unknown> 0x000000011761dae2
<unknown> 0x000000011761da00
<unknown> 0x000000011761da00
<unknown> 0x0000000117614849
JavaCalls::call_helper(JavaValue*, methodHandle const&, JavaCallArguments*, Thread*) 0x000000010bf3a582
InstanceKlass::call_class_initializer(Thread*) 0x000000010bf22af7
InstanceKlass::initialize_impl(Thread*) 0x000000010bf2244f
Reflection::invoke_constructor(oopDesc*, objArrayHandle, Thread*) 0x000000010c1ebdbb
JVM_NewInstanceFromConstructor 0x000000010bfc14f6
<unknown> 0x0000000117623950
<unknown> 0x000000011761da00
<unknown> 0x000000011761da00
<unknown> 0x000000011761dae2
<unknown> 0x000000011761da00
<unknown> 0x000000011761da00
<unknown> 0x000000011761dae2
<unknown> 0x000000011761dae2
<unknown> 0x000000011761dcec
<unknown> 0x000000011761da00
<unknown> 0x000000011761da00
<unknown> 0x000000011761dae2
<unknown> 0x000000011761da00
<unknown> 0x000000011761da00
<unknown> 0x000000011761dae2
<unknown> 0x000000011761da00
<unknown> 0x000000011761da00
<unknown> 0x000000011761da00
<unknown> 0x000000011761da00
<unknown> 0x000000011761da00
<unknown> 0x000000011761da00
<unknown> 0x0000000117614849
JavaCalls::call_helper(JavaValue*, methodHandle const&, JavaCallArguments*, Thread*) 0x000000010bf3a582
jni_invoke_static(JNIEnv_*, JavaValue*, _jobject*, JNICallType, _jmethodID*, JNI_ArgumentPusher*, Thread*) 0x000000010bf7e2af
jni_CallStaticVoidMethodV 0x000000010bf81c69
JNIEnv_::CallStaticVoidMethod(_jclass*, _jmethodID*, ...) jni.h:1521
main main.cpp:80
start 0x00007fff6f6563d5
start 0x00007fff6f6563d5
```

It doesn't stop in my code. I don't understand why this is happening, or if it's possible to avoid them, aside from ignoring all SIGBUSs.

I minimized my code and created the most simple example which reproduces the issue. Basically I created a cpp project which starts a jvm with **org/junit/platform/console/ConsoleLauncher** as main, with one simple test class. And the SIGBUS happens. It happens before my test even run. If I put my code as main instead of JUnit, it doesn't happen.

You can get a simple project which demonstrate the issue here: https://github.com/tallavi/sigbus-reproduction

Another thing I noticed. This example gives a single SIGBUS. If I put two test classes, I get two SIGBUSes, if I put 3, I get 3.

Thus, I suspect something within JUnit.

My environment:
MacOS 10.14.6
Clion 2019.3
Java 11.0.4 2019-07-16 LTS
JUnit versions I tried: 1.5.1, 1.5.2, 1.6.0-M1

Thanks!

**Comment by marcphilipp** [comment]
@tallavi I'm not sure how to help you here to be honest. My best guess from the stack is that it's caused by invoking the test class constructor visa reflection. If so, it should be reproducible without JUnit. Could you try writing a small programm with a `main()` method that does that?

**Comment by tallavi** [comment]
Hi @marcphilipp , thanks for the response. Since I wrote here, I had some responses in stack overflow https://stackoverflow.com/questions/59363909/diagnosing-sigbus-in-third-party-jars-used-with-jni-from-c, which made me improve my sample. You can pull the new one and see something more definitive.

Running the main in the java file runs the test without an issue. Running the same method from c++ does the test, but won't run the code after the tests are done.

From java:
```
/Library/Java/JavaVirtualMachines/jdk-11.0.2.jdk/Contents/Home/bin/java "-javaagent:/Applications/IntelliJ IDEA CE.app/Contents/lib/idea_rt.jar=59163:/Applications/IntelliJ IDEA CE.app/Contents/bin" -Dfile.encoding=UTF-8 -classpath /Users/tal/Development/v2x/qa-automation/sigbus-reproduction/out/production/classes:/Users/tal/.gradle/caches/modules-2/files-2.1/org.junit.platform/junit-platform-console-standalone/1.6.0-M1/7816f154e9de0286d1ce2b93af902fcff7b0e003/junit-platform-console-standalone-1.6.0-M1.jar FirstTest
withJUnit START
test START
test END

Thanks for using JUnit! Support its development at https://junit.org/sponsoring

╷
├─ JUnit Jupiter ✔
│  └─ FirstTest ✔
│     └─ myTest() ✔
└─ JUnit Vintage ✔

Test run finished after 114 ms
[         3 containers found      ]
[         0 containers skipped    ]
[         3 containers started    ]
[         0 containers aborted    ]
[         3 containers successful ]
[         0 containers failed     ]
[         1 tests found           ]
[         0 tests skipped         ]
[         1 tests started         ]
[         0 tests aborted         ]
[         1 tests successful      ]
[         0 tests failed          ]

withJUnit END

Process finished with exit code 0
```

Note the "withJUnit END" log line, which shows that the code continued to run after the test is done.

Same code, but now invoked from a jvm started from c++:
```
/Users/tal/Development/v2x/qa-automation/sigbus-reproduction/out/main
CppMainStart
current_path: /Users/tal/Development/v2x/qa-automation/sigbus-reproduction/out
Loading JAR: jars/junit-platform-console-standalone-1.5.2.jar
Loading JAR: jars/.DS_Store
Loading JAR: jars/junit-platform-console-standalone-1.6.0-M1.jar
Loading JAR: jars/sigbus-reproduction.jar
CreateVM:		JVM loaded successfully!
Before call
withJUnit START
test START
test END

Thanks for using JUnit! Support its development at https://junit.org/sponsoring

.
+-- JUnit Jupiter [OK]
| '-- FirstTest [OK]
|   '-- myTest() [OK]
'-- JUnit Vintage [OK]

Test run finished after 128 ms
[         3 containers found      ]
[         0 containers skipped    ]
[         3 containers started    ]
[         0 containers aborted    ]
[         3 containers successful ]
[         0 containers failed     ]
[         1 tests found           ]
[         0 tests skipped         ]
[         1 tests started         ]
[         0 tests aborted         ]
[         1 tests successful      ]
[         0 tests failed          ]


Process finished with exit code 0
```

Note that there's no "withJUnit END". After JUnit finishes, it somehow messed up the jvm or something?

Proof that the cpp invocation code works correctly, here's the same code, just calling a method which doesn't run junit:

```
/Users/tal/Development/v2x/qa-automation/sigbus-reproduction/out/main
CppMainStart
current_path: /Users/tal/Development/v2x/qa-automation/sigbus-reproduction/out
Loading JAR: jars/junit-platform-console-standalone-1.5.2.jar
Loading JAR: jars/.DS_Store
Loading JAR: jars/junit-platform-console-standalone-1.6.0-M1.jar
Loading JAR: jars/sigbus-reproduction.jar
CreateVM:		JVM loaded successfully!
Before call
withoutJUnit START
withoutJUnit END
After call
before destroying
after destroying
CppMainEnd

Process finished with exit code 0
```

Notice how the java side has finished, returned to the cpp side and finished running correctly.

Hope this information somehow helps. 

I'll be happy to provide more information, just ask.

Basically I'm out of ideas right now. The next thing to try is to run the same code but with the source code of console launcher instead of a compiled jar, perhaps then I'll see where in the source code everything breaks.

Thanks.

**Comment by marcphilipp** [comment]
Executing `./gradlew :junit-platform-console-standalone:publishToMavenLocal` publishes the JAR to the local Maven repository. I have no experience using JNI to create a JVM from C++. From the comments on the SO thread, it sounds like it's an issue with your environment or local libraries, isn't it?

**Comment by tallavi** [comment]
Hi @marcphilipp,

Ok, I managed to do that and actually step into the junit code from my code. I hoped that doing that  will somehow improve the enigmatic stack trace I got, but it did nothing. Regarding what Oo.oO wrote in SO, I don't see how it can possibly be an issue of some third party (other than JUnit), I isolated all the factors:

1. Running it directly from JAVA works, proving my JAVA code doesn't have issues.
2. Running the same code from c++ which calls the JAVA code, fails.
3. Running the exact same code from c++ which calls the JAVA, just omitting the JUnit part, works.

But I'm officially out of leads. Without building JDK from source and running it then, I don't think I'll get any further, and I don't think I'll do that..

**Comment by marcphilipp** [comment]
Do you need to call the Java code from C++ or could you instead launch a Java process from C++ as a workaround?

**Comment by marcphilipp** [comment]
@tallavi Any news on this? Is it still an issue for you or can we close this?

**Comment by tallavi** [comment]
Hi @marcphilipp,

Sorry, I missed your previous question. Launching a process is not an option because the entire purpose of this exercise is to have the java load a native library and debug it, so I need it in the same process.

Nothing new. I'm considering trying TestNg and see if it happens there too.. No other leads aside from that. For now, we ignore the signals.

**Comment by marcphilipp** [comment]
Ok, in that case, I'm closing this issue for now. Feel free to reopen if there's any new information.
